pyccel/parser/semantic.py

```diff
@@ -1358,47 +1392,65 @@ def _annotate_the_called_function_def(self, old_func, function_call_args=None):
         old_func : FunctionDef|Interface
            The function that needs to be annotated.
 
-        function_call_args : list[FunctionCallArgument], optional
+        function_call_args : list[FunctionCallArgument]
            The list of the call arguments.
 
         Returns
         -------
         func: FunctionDef|Interface
             The new annotated function.
         """
+        cls_base_syntactic = old_func.get_direct_user_nodes(lambda p: isinstance(p, ClassDef))
+        if cls_base_syntactic:
+            cls_name = cls_base_syntactic[0].name
+            cls_base = self.scope.find(cls_name, 'classes')
+            cls_scope = cls_base.scope
+            new_scope = cls_scope
+        else:
+            func_scope = old_func.scope if isinstance(old_func, FunctionDef) else old_func.syntactic_node.scope
+            new_scope = func_scope
         # The function call might be in a completely different scope from the FunctionDef
         # Store the current scope and go to the parent scope of the FunctionDef
-        old_scope            = self._scope
+        old_scope = self._scope
         old_current_function = self._current_function
-        names = []
-        sc = old_func.scope if isinstance(old_func, FunctionDef) else old_func.syntactic_node.scope
-        while sc.parent_scope is not None:
-            sc = sc.parent_scope
-            if not sc.name is None:
-                names.append(sc.name)
-        names.reverse()
-        if names:
-            self._current_function = DottedName(*names) if len(names)>1 else names[0]
-        else:
-            self._current_function = None
+        old_current_function_name = self._current_function_name
 
-        while names:
-            sc = sc.sons_scopes[names[0]]
-            names = names[1:]
+        # Walk up scope to root to find names of relevant scopes
+        scope_names = []
+        while new_scope.parent_scope is not None:
+            new_scope = new_scope.parent_scope
+            if not new_scope.name is None:
+                scope_names.append(new_scope.name)
+        scope_names.reverse()
+
+        # Use scope_names to find semantic scopes
+        while scope_names:
+            new_scope = new_scope.sons_scopes[scope_names[0]]
+            scope_names = scope_names[1:]
```

Comment:
I have the feeling that this code can be rewritten with no reverse() and no redefining of the list in the loop.

Reply:
Fixed [10683bf](https://github.com/pyccel/pyccel/pull/2314/commits/10683bfdb7ea47803f29a4d2634b3e32618fb29e)